src/index.ts

```diff
@@ -0,0 +1,322 @@
+import {
+  Environment,
+  Platform,
+  platformFromString,
+  DismissBlockingDialog,
+  SNAlertService,
+  SNApplication,
+  DeviceInterface,
+  EncryptionIntent,
+  StorageEncryptionPolicies
+} from 'snjs';
+import { SNWebCrypto } from 'sncrypto/dist/sncrypto-web';
+import { BackupFile } from 'snjs/dist/@types/services/protocol_service';
+
+declare const JSZip: any;
+
+const KEYCHAIN_STORAGE_KEY = 'keychain';
+
+class WebDeviceInterface extends DeviceInterface {
+  constructor() {
+    super(window.setTimeout, window.setInterval);
+  }
+
+  openUrl(): void {
+    throw new Error('Method not implemented.');
+  }
+
+  async getRawStorageValue(key: string) {
+    return localStorage.getItem(key);
+  }
+
+  async getAllRawStorageKeyValues() {
+    return Object.entries(localStorage).map(([key, value]) => ({
+      key,
+      value,
+    }));
+  }
+
+  async setRawStorageValue(key: string, value: string) {
+    localStorage.setItem(key, value);
+  }
+
+  async removeRawStorageValue(key: string) {
+    localStorage.removeItem(key);
+  }
+
+  async removeAllRawStorageValues() {
+    localStorage.clear();
+  }
+
+  async openDatabase(_identifier: string) {
+    return {};
+  }
+
+  _getDatabaseKeyPrefix(identifier: any) {
+    if (identifier) {
+      return `${identifier}-item-`;
+    } else {
+      return 'item-';
+    }
+  }
+
+  _keyForPayloadId(id: any, identifier: any) {
+    return `${this._getDatabaseKeyPrefix(identifier)}${id}`;
+  }
+
+  async getAllRawDatabasePayloads(identifier: any) {
+    const models = [];
+    for (const key in localStorage) {
+      if (key.startsWith(this._getDatabaseKeyPrefix(identifier))) {
+        models.push(JSON.parse(localStorage[key]));
+      }
+    }
+    return models;
+  }
+
+  async saveRawDatabasePayload(payload: { uuid: any }, identifier: any) {
+    localStorage.setItem(
+      this._keyForPayloadId(payload.uuid, identifier),
+      JSON.stringify(payload)
+    );
+  }
+
+  async saveRawDatabasePayloads(payloads: any, identifier: any) {
+    for (const payload of payloads) {
+      await this.saveRawDatabasePayload(payload, identifier);
+    }
+  }
+
+  async removeRawDatabasePayloadWithId(id: any, identifier: any) {
+    localStorage.removeItem(this._keyForPayloadId(id, identifier));
+  }
+
+  async removeAllRawDatabasePayloads(identifier: any) {
+    for (const key in localStorage) {
+      if (key.startsWith(this._getDatabaseKeyPrefix(identifier))) {
+        delete localStorage[key];
+      }
+    }
+  }
+
+  /** @keychain */
+  async getNamespacedKeychainValue(identifier: string | number) {
+    const keychain = await this.getRawKeychainValue();
+    if (!keychain) {
+      return;
+    }
+    return keychain[identifier];
+  }
+
+  async setNamespacedKeychainValue(value: any, identifier: any) {
+    let keychain = await this.getRawKeychainValue();
+    if (!keychain) {
+      keychain = {};
+    }
+    localStorage.setItem(
```

Comment:
This device interface really should be memory based overall. It's another insurance policy against data persistence. So the keychain should be an in-memory data structure, as well as the payload store. Really we shouldn't see any references of localStorage.